packages/dds/tree/src/shared-tree/defaultCommitEnricher.ts

```diff
@@ -0,0 +1,176 @@
+	 * @param revision - the revision associated with the change.
+	 * @returns the enriched change. Possibly the same as the one passed in.
+	 */
+	updateChangeEnrichments(change: TChange, revision: RevisionTag): TChange;
+
+	/**
+	 * Applies a change to the tip state.
+	 * @param change - the change to apply. Not mutated.
+	 * @param revision - the revision associated with the change.
+	 * Can be undefined when the applied change is a rollback.
+	 */
+	applyTipChange(change: TChange, revision?: RevisionTag): void;
+
+	/**
+	 * Disposes of the enricher.
+	 */
+	[disposeSymbol](): void;
+}
+
+/**
+ * Default implementation of {@link ICommitEnricher}.
+ */
+export class DefaultCommitEnricher<TChange> implements ICommitEnricher<TChange> {
+	/**
+	 * Change enricher that represent the tip of the top-level local branch (i.e., the branch on which in-flight
+	 * commits are applied and automatically rebased).
+	 */
+	private tip: ChangeEnricherCheckout<TChange>;
+
+	/**
+	 * The list of commits (from oldest to most recent) that have been submitted but not sequenced.
+	 */
+	private inFlightQueue: GraphCommit<TChange>[] = [];
+
+	/**
+	 * The list of commits (from oldest to most recent) that should be resubmitted.
+	 */
+	private resubmitQueue: GraphCommit<TChange>[] = [];
+
+	/**
+	 * Represents the index in the `inFlight` array of the most recent in flight commit that has
+	 * undergone rebasing but whose enrichments have not been updated.
+	 * All in-flight commits with an index inferior or equal to this number have stale enrichments.
+	 *
+	 * Is -1 when *any* of the following is true:
+	 * - There are no in-flight commits (i.e., no local commits have been made or they have all been sequenced)
+	 * - None of the in-flight commits have been rebased
+	 * - In-flight commits that have been rebased have all had their enrichments updated
+	 */
+	private latestInFlightCommitWithStaleEnrichments: number = -1;
+
+	public constructor(
+		/**
+		 * A function that can invert a change.
+		 */
+		private readonly inverter: ChangeRebaser<TChange>["invert"],
+		/**
+		 * A factory that return a new {@link ChangeEnricherCheckout} that corresponds to the tip of the top-most local
+		 * branch.
+		 */
+		private readonly checkoutFactory: () => ChangeEnricherCheckout<TChange>,
+	) {
+		this.tip = this.checkoutFactory();
+	}
+
+	public enrichCommit(commit: GraphCommit<TChange>): GraphCommit<TChange> {
+		const toResubmit = this.resubmitQueue.shift();
+		if (toResubmit !== undefined) {
+			this.inFlightQueue.push(toResubmit);
+			return toResubmit;
+		} else {
+			const updatedChange = this.tip.updateChangeEnrichments(commit.change, commit.revision);
+			const updatedCommit = { ...commit, change: updatedChange };
+			this.inFlightQueue.push(updatedCommit);
+			this.tip[disposeSymbol]();
+			this.tip = this.checkoutFactory();
+			return updatedCommit;
+		}
+	}
+
+	public prepareForResubmit(toResubmit: readonly GraphCommit<TChange>[]): void {
+		assert(
```

Comment:
Do you also want to assert that `toResubmit.length > 0`? I think that is always true in practice. Or maybe that's not a requirement - but if `toResubmit.length` _is_ 0, then this function can early out, so it doesn't waste work? Or maybe it's fine as is, since that never happens.

Reply:
Since we don't expect it to ever happen but we would handle it correctly if it did, I'm reluctant to pay the code and perf code of having it. It's such a minimal cost though, I really don't mind. Do you have a preferrence?

Reply:
I'd say just leave it as-is if it's a non-possibility for the foreseeable future.